### PR TITLE
Added workshop that introduces trainees to the basics of TypeScript

### PR DESCRIPTION
**Pull request by StevenVanBlerk**
# Intention
This workshop is intended for trainees with some basic understanding of typed languages but with no experience with TypeScript. It will require volunteers having familiarity with TypeScript, but strong familiarity with another typed language might be sufficient.

This workshop was originally built for and used within [SDC Sprint 4](https://sdc.codeyourfuture.io/decomposition/sprints/4/) (Types and typechecking) as a day plan workshop. This PR includes a `DAY_PLAN_WORKSHOP.md` file that could be transposed into [the day plan](https://sdc.codeyourfuture.io/decomposition/sprints/4/day-plan/) if some minor changes are made to match the markdown of the day plans.

If you'd like more context on how this workshop came about, there is [an extensive Slack thread](https://codeyourfuture.slack.com/archives/C0884D7GLS3/p1754136156817339) discussing it.

# Things to keep in mind while reviewing this
## Is the write up sensible?
This is the 2nd iteration (3rd of 4th in my head) of this workshop. Please keep an eye out for text that feels out of place. I've proof read everything but I might have left bits that made more sense in previous iterations.

## Does this code run well?
I'd appreciate reviewers running and completing the exercises themselves if possible. The config became a bit messy due to different exercises requiring different testing patterns. I think the tests are giving helpful feedback to trainees but there might be edge cases where unexpected errors come through. 

Exercise 2 utilises experimental vitest features to test at compile time. This exercise is the most likely to result in unexpected or unhelpful behaviour.

# Future improvements and my thoughts on further TypeScript exposure
This workshop aimed to automate as much of the grading of exercises as possible. I used vitest to accomplish this, which proved quite tricky since vitest is mostly intended for testing at runtime. This means it is difficult to ask a trainee to build a `type SomeType` and then automate a test against `SomeType`. 

If another means of testing code at compile time can be found, it might be viable to follow the approach of this workshop. Otherwise it might be best to take a different approach for further TypeScript work. Having volunteers more involved in the grading and feedback, for example how `exercise3.ts` operates, might be viable.

My recommendation would be to have further learning of TypeScript be a byproduct and not the main focus. This could be accomplished by having JavaScript backlog work keep its current content but rather be written in TypeScript.

## How I've seen others teach TypeScript
[Total TypeScript](https://www.totaltypescript.com/tutorials/beginners-typescript) has an excellent and short beginner course, which was the main inspiration behind these exercises. I'm guessing they encountered the same problems as me, hence their later exercises also not leaning on automated tests.

[W3Schools](https://www.w3schools.com/typescript/typescript_exercises.php) takes a different approach by focusing on knowledge of TypeScript. Their exercises are not run within an IDE and tend to be very short code snippets.
image

## An idea for an extra exercise
Should someone have capacity to further this workshop, I think it would be beneficial to trainees to have an exercise that highlights the benefits of returning to old code written in TypeScript instead of JavaScript. As Daniel poignantly put it:

> One of the really useful things about static types that our trainees typically don't understand is that they help when we change code. Our trainees typically only think about writing code for the first time.
> 
> But showing them "If we rename Piscine to Assessment, or if we delete ITD, or if Users now need a phone number, TS will tell us everywhere that we need to change our code" is really useful.
> 
> I'm not sure the best way to introduce this - maybe we can introduce an exercise where they make a change to existing code and see that it was useful? (This would re-enforce the teaching in https://sdc.codeyourfuture.io/tools/sprints/5/prep/#type-guided-refactorings with a practical exercise, and show how TS helps in the same way MyPY does)

### CODE REVIEW

**Review by Poonam-raj — CHANGES_REQUESTED** [review]
The purpose of this workshop is clearly filling a gap in knowledge. I've come at this review in the mindset of a fresh volunteer or trainee. 

Based on the material around types it could be plausible that some students have never engaged with the type syntax properly before and with this workshop being an introduction it feels apt to cover the foundations so a few of my suggestions encourage breaking down the basic syntax.

A few other suggestions touch on phrasing. And there's a bit more about how the Learning Objectives connect with the workshop material. 

Few general suggestions:
- Re: phrasing - mainly my aim was to bring the facts to the front and the context second. Looking to be as clear in language as possible.
- Generally more instruction is needed in how this workshop is executed, to enable volunteers and trainees to know exactly what their role in each section will be.

Please let me know if anything is unclear or if anything actually doesn't fit the intention for the workshop.
Thanks!

File: introduction-to-typescript/DAY_PLAN_WORKSHOP.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+- Breakout into groups of about 3 to 5 trainees.
+- Make sure you have access to the [exercise repository](https://github.com/CodeYourFuture/CYF-Workshops/tree/main/introduction-to-typescript).
+- Clone it to a local project.
+- Follow the instructions in `README.md` to setup the project and run the exercise.
+
+### Fixing errors
+
+Now that you've got the project running locally, your job is to ensure all tests are passing. There are TypeScript type examples within `src/examples` should you need a hint but feel free to also do your own research.
+
+## Takeaways (10 minutes)
```
[the remark this answers is not in the document]
Comment (StevenVanBlerk) [reply introduction-to-typescript/DAY_PLAN_WORKSHOP.md]: I was thinking for "Takeaways" to be any willing trainee sharing their thoughts with the entire class. I'll better emphasise this 👍

File: introduction-to-typescript/src/exercise1.ts
Comment (StevenVanBlerk) [inline introduction-to-typescript/src/exercise1.ts]: The aim of this exercise is to apply what they learnt during the prior `DAY_PLAN_WORKSHOP.md` content. The solution is quite straightforward and is essentially answered within `DAY_PLAN_WORKSHOP.md`.

I don't expect this exercise to take more than a few minutes for trainees but it does make their first hands-on experience with TypeScript manageable.

Comment (StevenVanBlerk) [reply introduction-to-typescript/src/exercise1.ts]: The simplest solution is to just have the function return an object that is of type `User`

Image

File: introduction-to-typescript/src/exercise2.ts
[the remark this answers is not in the document]
Comment (StevenVanBlerk) [reply introduction-to-typescript/src/exercise2.ts]: The simplest solution is to specify `a: number` and `b: number`, which causes the return type to be inferred as `number`

Image

**Review by Poonam-raj — COMMENTED** [review]
Thanks for addressing all my points, it reads so much clearer for both students and volunteers. I added one response to the group of 3-5 comment, mainly noting if we need to be explicit about working on the same fork. 

The other big question on my mind is whether there's so much new material here that this should be split between prep material (lines 1 - 179ish) and a workshop (the rest of the day plan). Unsure if this is a future iteration of the workshop and not something we edit now. Will ask on the thread

File: introduction-to-typescript/src/exercise3.ts
[the remark this answers is not in the document]
Comment (illicitonion) [reply introduction-to-typescript/src/exercise3.ts]: I think this is a really important narrative to present to volunteers running this workshop - maybe we can embed it in a `<details>` block in the exercises part of the `DAY_PLAN_WORKSHOP`?

**Review by illicitonion — COMMENTED** [review]
This looks great, thank you so much for putting it together!

File: introduction-to-typescript/src/exercise2.ts
[the remark this answers is not in the document]
Comment (StevenVanBlerk) [reply introduction-to-typescript/src/exercise2.ts]: I interestingly prefer inferred types wherever possible because I find explicit types risky 😆 I don't know what is more common either. My anecdotal experience has been that explicit types are more likely to mask unexpected types at runtime than inferred types but maybe that's just a byproduct of the projects I've worked on and not representative of TS as a whole.

TS philosophy aside, I do agree that an explicit solution to this problem is more 'educational' to trainees but I couldn't find a way of enforcing explicit types via tests.

[the remark this answers is not in the document]
Comment (StevenVanBlerk) [reply introduction-to-typescript/src/exercise2.ts]: I suppose W3 schools approach (see PR description) to teaching TS would allow for tasking trainees with specifying explicit return types 
image

**Review by Poonam-raj — APPROVED** [review]
When it comes to this workshop's content and clarity I'd say this is fab. I can't see any holes people might fall in, everything is nicely laid out for students and volunteers. In terms of preferences on TS use I have less to say there, but happy with the tradeoffs and priorities you've mentioned in the three exercises. Seems like a good playground for people to explore and form opinions too.
